src/idl_gen_general.cpp

```diff
@@ -1130,6 +1132,71 @@ void GenStruct(StructDef &struct_def, std::string *code_ptr) {
       code += "); }\n";
     }
   }
+  if (struct_def.has_key) {
+    bool is_string = key_field->value.type.base_type == BASE_TYPE_STRING;
+    std::string key_name = lang_.language == IDLOptions::kCSharp ? MakeCamel(key_field->name) : key_field->name + "()";
+    std::string key_type = GenTypeGet(key_field->value.type);
+    std::string key_offset = NumToString(key_field->value.offset);
+    std::string key_getter = is_string ? GenGetter(key_field->value.type) : GenGetter(key_field->value.type).substr(2);
+    std::string data_buffer = lang_.language == IDLOptions::kCSharp ? "builder.DataBuffer" : "bb";
+
+    code += "\n  public static int " + FunctionStart('K') + "eysCompare(" + key_type + " o1, " + key_type + " o2) ";
+    if (lang_.language == IDLOptions::kJava && !is_string) {
+      code += "{\n    if (o1 < o2) return -1;\n";
+      code += "    else if (o1 == o2) return 0;\n";
+      code += "    else return 1;\n  }\n";
+    }
+    else {
+      code += "{ return o1." + FunctionStart('C') + "ompareTo(o2); }\n";
+    }
+
+    code += "\n  public int " + FunctionStart('K') + "eyCompareWithValue(" + key_type + " val) ";
+    if (lang_.language == IDLOptions::kJava && !is_string) {
+      code += "{\n    if (" + key_name + " < val) return -1;\n";
+      code += "    else if (" + key_name + " == val) return 0;\n";
+      code += "    else return 1;\n  }\n";
+    }
+    else {
+      code += "{ return " + key_name + "." + FunctionStart('C') + "ompareTo(val); }\n";
+    }
+
+    code += "\n  public static ";
+    code += (lang_.language == IDLOptions::kCSharp ? "VectorOffset " : "int ") + FunctionStart('C') + "reateMySortedTableVector(FlatBufferBuilder builder, ";
```

Comment:
There's a LOT of language conditionals in this code, making this hard to read.
Quite a few can be added to the language parameters struct at the top of this file. Some that cannot, and that are used frequently, could be made into a function.